lib/WebService/SendBird.pm

```diff
@@ -0,0 +1,306 @@
+package WebService::SendBird;
+
+use strict;
+use warnings;
+
+use Mojo::UserAgent;
+use Mojo::URL;
+
+use WebService::SendBird::User;
+use WebService::SendBird::GroupChat;
+
+use Carp qw();
+
+our $VERSION = '0.001';
+
+# ABSTRACT: client to work with SendBird API
+
+=head1 NAME
+
+WebService::SendBird - unofficial support for the Sendbird API
+
+=head1 SYNOPSIS
+
+ use WebService::SendBird;
+
+ my $api = WebService::SendBird->new(
+     api_token => 'You_Api_Token_Here',
+     app_id    => 'You_App_ID_Here',
+ );
+
+ my $user = $api->create_user(
+     user_id     => 'my_chat_user_1',
+     nickname    => 'pumpkin',
+     profile_url => undef,
+ );
+
+ my $chat = $api->create_group_chat(
+     user_ids => [ $user->user_id ],
+ );
+
+=head1 DESCRIPTION
+
+Basic implementation for SendBird Platform API client, which helps to create users and group chats.
+
+More information at L<Platform API Documentation|https://docs.sendbird.com/platform>
+
+=cut
+
+use constant DEFAULT_API_URL_TEMPLATE => 'https://api-%s.sendbird.com/v3';
+use constant DEFAULT_REQUEST_TIMEOUT  => 15;
+
+=head1 METHODS
+
+=head2 new
+
+Creates an instance of API client
+
+=over 4
+
+=item * C<api_token> - Master or Secondary API Token.
+
+=item * C<app_id> - Application ID.
+
+=item * C<api_url> - URL to API end point. By default it will be generated from app_id.
+
+=item * C<ua> - Custom http client for API requests, should have the same interface like L<Mojo::UserAgent>.
+
+=item * C<timeout> - request timeout, default value 15 seconds
+
+=back
+
+=cut
+
+
+sub new {
+    my ($cls, %params) = @_;
+
+    Carp::croak('Missing required argument: api_token') unless $params{api_token};
+    Carp::croak('Missing required argument: app_id or api_url') unless $params{app_id} || $params{api_url};
+
+    my $self = +{
+        api_token => $params{api_token},
+        $params{ua}      ? (ua      => $params{ua})                      : (),
+        $params{app_id}  ? (app_id  => $params{app_id})                  : (),
+        $params{api_url} ? (api_url => Mojo::URL->new($params{api_url})) : (),
+    };
+
+    return bless $self, $cls;
+}
+
+
+=head2 app_id
+
+Returns Application ID.
+
+=cut
+
+sub app_id { shift->{app_id} }
+
+=head2 api_token
+
+Returns API Token
+
+=cut
+
+sub api_token { shift->{api_token} }
+
+=head2 api_url
+
+Returns API endpoint url
+
+=cut
+
+sub api_url {
+    my $self = shift;
+
+    $self->{api_url} //= Mojo::URL->new(sprintf(DEFAULT_API_URL_TEMPLATE, $self->app_id));
+
+    return $self->{api_url};
+}
+
+=head2 timeout
+
+Return http request timeout value.
+
+=cut
+
+sub timeout { shift->{timeout} }
+
+=head2 ua
+
+Return User Agent for http request.
+
+=cut
+
+sub ua {
+    my $self = shift;
+
+    return $self->{ua} if $self->{ua};
+
+    my $ua = Mojo::UserAgent->new();
+
+    $ua->inactivity_timeout($self->timeout || DEFAULT_REQUEST_TIMEOUT);
+    $ua->proxy->detect;
+    $ua->max_connections(100);
```

Comment:
Why we need more than 1?

Reply:
yes. you're right it's sync code we will have only one connection

Reply:
fixed